File: valid-anagram.cpp
#include <iostream>
#include <map>

using namespace std;

bool isAnagram(string s, string t) {
  map<char, int> letters;

  if (s.length() != t.length())
    return false;
  for (auto c: s) {
    letters[c]++;
  }

  for (auto c: t) {
    letters[c]--;
    if (letters[c] < 0)
      return false;
  }

  return true;
}

int main(int argc, char const *argv[]) {

  cout << isAnagram("anagram", "nagaram") << "\n";
  cout << isAnagram("rat", "car") << "\n";

  return 0;
}

File: valid-anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False
        if s == t:
            return True
        # dictionaries for s and t.
        # Holds the number of occurrences of each character in the string.
        countS, countT = {}, {}

        for c in s:
            countS[c] = 1 + countS.get(c, 0)

        for c in t:
            countT[c] = 1 + countT.get(c, 0)

        for c in countS.keys():
            if countS[c] != countT.get(c, 0):
                return False
        return True


s = Solution()
result = s.isAnagram("anagram", "nagaram")
print(result)
